pynicotine/shares.py

```diff
@@ -834,15 +858,120 @@ def process_scanner_messages(self, scanner, scanner_queue):
 
         return False
 
-    def rescan_shares(self, init=False, rescan=True, rebuild=False, use_thread=True):
+    def count_shares(self, shares=None, group=None, logging=True):
+        """ Check if any shares are missing and log counts """
+
+        reads, fails = self.check_shares(shares=shares, group=group)
+
+        num_reads = len(reads)
+        num_fails = len(fails)
+        num_total = num_reads + num_fails
+
+        summary_total = _(f"{num_total} shares configured")
+        summary_reads = _(f"{num_reads} shares ready")
+        summary_fails = _(f"{num_fails} shares not found")
+
+        reads_text = '\n\n'.join(reads)
+        fails_text = '\n\n'.join(fails)
+
+        if logging:
+            log.add(summary_total)
+
+            if num_total:
+                log.add(summary_reads)
+                if num_fails:
+                    log.add(summary_fails)
+            else:
+                fails.insert(0, summary_total)
+
+            if reads:
+                log.add_transfer(f"{summary_reads}:\n\n{reads_text}\n")
+
+            if fails and num_total:
+                log.add_transfer(f"{summary_fails}:\n\n{fails_text}\n")
+
+            log.add_transfer(_(f"{summary_reads} out of {summary_total}"))
+
+        return (num_reads, num_fails, num_total, summary_total,
+                summary_reads, summary_fails, reads_text, fails_text, fails)
+
+    def check_shares(self, shares=None, group=None):
+        """ Returns TWO lists of strings: readable shares, and missing shares.
+        group specifies which shares to check: "normal", "buddy", etc. """
+
+        if shares is None:
+            # Triggered by list shares command
+            shares = self.get_shared_folders()
+
+        group_index = self.get_group_index(group) if group is not None else 0
+        reads, fails = [], []
+
+        for table in shares:
+            group_name = self.get_group_name(group_index).title()
+
+            for virtual, folder, *_unused in table:
+                if os.access(folder, os.R_OK):
+                    reads.append(_(f"{group_name} share \"{virtual}\" ready at:\n{folder}"))
+                else:
+                    fails.append(_(f"{group_name} share \"{virtual}\" not found at:\n{folder}"))
+
+            if group is not None:
+                # Only check a specific group (ie just "normal" or "buddy" shares, etc)
+                break
+
+            group_index += 1
+
+        return reads, fails
+
+    def confirm_force_rescan(self, shares):
+
+        # Check if any shares are missing and log counts
+        (num_reads, num_fails, num_total, summary_total,
+         summary_reads, summary_fails, _reads_text, fails_text, fails) = self.count_shares(shares)
+
+        if num_reads and not num_fails:
+            # Continue initializing shares, and do the rescan now
+            return True
+
+        log.add(_("Rescan aborted") + ": " + (fails_text.replace(":\n", " ") if num_fails == 1 else summary_fails))
```

Comment:
In the future, please avoid creating strings like this. It becomes difficult to understand what the message is, without jumping back and forth in the code to see what fails_text is.

We also can't be sure that every language wants a colon that looks exactly like this at this specific location (French uses a space before punctuation, for example).

Reply:
Indeed it is the product of an over-complicated design. I learn more from doing this project that these over-engineered string hacks are hardly ever suitable for all cases.

That's interesting I did not know about the French language punctuation.

Reply:
It takes a bit of time to get used to, but the more you work on different projects, the more you appreciate a simple solution that does just enough, over of a complex solution with all kinds of bells and whistles. All about finding a balance that both the developer/maintainer and user appreciates.

Rule of thumb for translated strings is to always try to use full sentences.